src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala

```diff
@@ -75,49 +130,120 @@ object RockSlicer {
+              else unique
+            }
+          if (allFaces.diff(nonSharedFaces).isEmpty) {
+            Block(origin, nonSharedFaces)
+          }
+        }
+      }
+    }).collect { case blockType: Block => blockType }
+
+    // Remove redundant faces and remove duplicates
+    val nonRedundantBlocks = reconstructedBlocks map { case block @ Block(center, _) =>
+      Block(center, block.nonRedundantFaces)
+    }
+    val joinedBlocks = 
+      nonRedundantBlocks.foldLeft(Seq.empty[Block]) { (unique, current) =>
+        if (!unique.exists(Block.compareBlocks(current, _)))
+          current +: unique
+        else unique
+      }
+
+    // Divide blocks into groups with and without processor joints remaining
+    val (remainingBlocks, completedBlocks) = joinedBlocks.partition { block =>
+      block.faces.exists(_.processorJoint)
+    }
+
+    if (remainingBlocks.nonEmpty) {
+      // Merged blocks still contain some processor joints
+      mergeBlocks(remainingBlocks ++ processorBlocks.tail, completedBlocks ++ mergedBlocks, origin)
+    } else if (processorBlocks.tail.isEmpty) {
+      // All blocks are free of processor joints - check for duplicates then return
+      val mergedBlocksDuplicates = completedBlocks ++ mergedBlocks
+      mergedBlocksDuplicates.foldLeft(Seq[Block]()) { (unique, current) =>
+        if (!unique.exists(Block.compareBlocks(current, _)))
+          current +: unique
+        else
+          unique
+      }
+    } else {
+      // Proceed to next processor block
+      mergeBlocks(processorBlocks.tail, completedBlocks ++ mergedBlocks, origin)
     }
   }
 
-  private def loadBalanceIndices(indices: Seq[Int]): Seq[Int] = {
-    def loadBalanceIndicesImpl(queue: mutable.PriorityQueue[Seq[Int]]): Seq[Int] = {
-      if (queue.isEmpty) {
-        Nil
-      } else {
-        val s = queue.dequeue()
-        val (median, left, right) = removeMedian(s)
-        if (left.nonEmpty) {
-          queue.enqueue(left)
-        }
-        if (right.nonEmpty) {
-          queue.enqueue(right)
+  /**
+    * Compares two input blocks and determines whether they share a processor face
+    * @param block1 First input block
+    * @param block2 Second input block
+    * @return List of processor faces that are shared by the two blocks. Will be 
+    *         empty if they share no faces
+    */
+  def compareProcessorBlocks(block1: Block, block2: Block): Seq[Face] = {
+    val processorFaces1 = block1.faces.filter { case face => face.processorJoint }
+    val processorFaces2 = block2.faces.filter { case face => face.processorJoint }
+    val faceMatches = 
+      processorFaces1 map { case face1 =>
+        processorFaces2 map { case face2 =>
+          if (compareSharedFaces(face1, face2)) {
+            Seq[Face](face1, face2)
+          } else Seq[Face]()
         }
-        median +: loadBalanceIndicesImpl(queue)
       }
-    }
+    faceMatches.filter{ case faces => faces.nonEmpty }.flatten.flatten.distinct
```

Comment:
You can just do `faceMatches.filter(_.nonEmpty)` here, right? You don't have to use the whole `case` syntax in some simple situations, and in fact you don't really need `case` unless you want to destructure each element, so `faceMatches.filter { faces => faces.nonEmpty }` should work here.